stdlib/ephemeron.ml

```diff
@@ -0,0 +1,553 @@
+        | Cons(_,_,next) -> replace_bucket next
+      in
+      let i = key_index h hkey in
+      let l = h.data.(i) in
+      try
+        replace_bucket l
+      with Not_found ->
+        let container = H.create key info in
+        h.data.(i) <- Cons(hkey, container, l);
+        h.size <- h.size + 1;
+        if h.size > Array.length h.data lsl 1 then resize key_index h
+
+    let mem h key =
+      let hkey = H.hash h.seed key in
+      let rec mem_in_bucket = function
+      | Empty ->
+          false
+      | Cons(hk, c, rest) when hk = hkey ->
+          begin match H.equal key c with
+          | ETrue -> true
+          | EFalse -> mem_in_bucket rest
+          | EDead ->
+              (** TODO? remove this dead key *)
+              mem_in_bucket rest
+          end
+      | Cons(hk, c, rest) -> mem_in_bucket rest in
+      mem_in_bucket h.data.(key_index h hkey)
+
+    let iter f h =
+      let rec do_bucket = function
+        | Empty ->
+            ()
+        | Cons(_, c, rest) ->
+            begin match H.get_key c, H.get_data c with
+            | None, _ | _, None -> (** TODO? remove this dead key? *) ()
+            | Some k, Some d -> f k d
+            end; do_bucket rest in
+      let d = h.data in
+      for i = 0 to Array.length d - 1 do
+        do_bucket d.(i)
+      done
+
+    let fold f h init =
+      let rec do_bucket b accu =
+        match b with
+          Empty ->
+            accu
+        | Cons(_, c, rest) ->
+            let accu = begin match H.get_key c, H.get_data c with
+              | None, _ | _, None -> (** TODO? remove this dead key? *) accu
+              | Some k, Some d -> f k d accu
+            end in
+            do_bucket rest accu  in
+      let d = h.data in
+      let accu = ref init in
+      for i = 0 to Array.length d - 1 do
+        accu := do_bucket d.(i) !accu
+      done;
+      !accu
+
+    let length h = h.size
+
+
+    let rec bucket_length accu = function
+      | Empty -> accu
+      | Cons(_, _, rest) -> bucket_length (accu + 1) rest
+
+    let stats h =
+      let mbl =
+        Array.fold_left (fun m b -> max m (bucket_length 0 b)) 0 h.data in
+      let histo = Array.make (mbl + 1) 0 in
+      Array.iter
+        (fun b ->
+           let l = bucket_length 0 b in
+           histo.(l) <- histo.(l) + 1)
+        h.data;
+      { Hashtbl.num_bindings = h.size;
+        num_buckets = Array.length h.data;
+        max_bucket_length = mbl;
+        bucket_histogram = histo }
+
+    let rec bucket_length_alive accu = function
+      | Empty -> accu
+      | Cons(_, c, rest) when H.check_key c ->
+          bucket_length_alive (accu + 1) rest
+      | Cons(_, _, rest) -> bucket_length_alive accu rest
+
+    let stats_alive h =
+      let size = ref 0 in
+      let mbl =
+        Array.fold_left (fun m b -> max m (bucket_length 0 b)) 0 h.data in
+      let histo = Array.make (mbl + 1) 0 in
+      Array.iter
+        (fun b ->
+           let l = bucket_length 0 b in
```

Comment:
In stats_alive: Why is the number of bindings in bucket b counted via the bucket_length function?
Is the function expected to compute the number of bindings that are still alive? In that case, should the call to bucket_length be replaced by a call of bucket_length_alive?
I expect this would produce the actual number of bindings that are still alive within the bucket?